runhouse/resources/hardware/utils.py

```diff
@@ -248,3 +258,152 @@ def up_cluster_helper(cluster, capture_output: Union[bool, str] = True):
+        headers=rns_client.request_headers(),
+    )
+
+    return clusters_in_den_resp
+
+
+def get_unsaved_live_clusters(den_clusters: list[Dict], sky_live_clusters: list):
+    den_clusters_names = [c.get("name") for c in den_clusters]
+
+    # getting the on-demand clusters that are not saved in den.
+    if sky_live_clusters:
+        return [
+            cluster
+            for cluster in sky_live_clusters
+            if f'/{rns_client.username}/{cluster.get("Name")}' not in den_clusters_names
+        ]
+    else:
+        return []
+
+
+def get_running_and_not_running_clusters(clusters: list):
+    running_clusters, not_running_clusters = [], []
+
+    for den_cluster in clusters:
+        # get just name, not full rns address. reset is used so the name will be printed all in white.
+        cluster_name = f'[reset]{den_cluster.get("name").split("/")[-1]}'
+        cluster_type = den_cluster.get("data").get("resource_subtype")
+        cluster_status = (
+            den_cluster.get("status") if den_cluster.get("status") else "unknown"
+        )
+
+        # currently relying on status pings to den as a sign of cluster activity.
+        # The split is required to remove milliseconds and the offset (according to UTC) from the timestamp.
+        # (status_last_checked is in the following format: YYYY-MM-DD HH:MM:SS.ssssss±HH:MM)
+
+        last_active_at = den_cluster.get("status_last_checked")
+        last_active_at = (
+            datetime.datetime.fromisoformat(last_active_at.split(".")[0])
+            if isinstance(last_active_at, str)
+            else datetime.datetime(1970, 1, 1)
+        )  # Convert to datetime
+        last_active_at = last_active_at.replace(tzinfo=datetime.timezone.utc)
+
+        if cluster_status == "running" and not last_active_at:
+            # For BC, in case there are clusters that were saved and created before we introduced sending cluster status to den.
+            cluster_status = "unknown"
+
+        cluster_info = {
+            "Name": cluster_name,
+            "Cluster Type": cluster_type,
+            "Status": cluster_status,
+            "Last Active (UTC)": last_active_at,
+        }
+        running_clusters.append(
+            cluster_info
+        ) if cluster_status == "running" else not_running_clusters.append(cluster_info)
+
+    # Sort clusters by the 'Last Active (UTC)' and 'Status' column
+    not_running_clusters = sorted(
+        not_running_clusters,
+        key=lambda x: (x["Last Active (UTC)"], x["Status"]),
+        reverse=True,
+    )
+    for cluster in not_running_clusters:
+        cluster["Last Active (UTC)"] = cluster.get("Last Active (UTC)").strftime(
+            "%m/%d/%Y, %H:%M:%S"
+        )
+
+    running_clusters = sorted(
+        running_clusters, key=lambda x: x["Last Active (UTC)"], reverse=True
+    )
+
+    for cluster in running_clusters:
+        cluster["Last Active (UTC)"] = cluster.get("Last Active (UTC)").strftime(
```

Comment:
nope, because we pass only den clusters to this function (`get_running_and_not_running_clusters`), and den clusters will always have a "Last Active (UTC)" field.